react/ProductReference.tsx

```diff
@@ -0,0 +1,68 @@
+import React, { FunctionComponent, useState, useEffect } from 'react'
+import { Loading } from 'vtex.render-runtime'
+import { useCssHandles } from 'vtex.css-handles'
+
+import { useItemContext } from './ItemContext'
+import { opaque } from './utils/opaque'
+import {
+  PRODUCT_REFERENCE_ID,
+  PRODUCT_SKU_REFERENCE_ID,
+  PRODUCT_ID,
+  PRODUCT_SKU_ITEM_ID,
+} from './constants/Identifiers'
+
+const CSS_HANDLES = [
+  'productIdentifier',
+  'productIdentifierValue',
+  'productIdentifierLabelValue',
+] as const
+
+interface Props {
+  identifierLabel?: string
+  identifierOption?: string
+}
+
+const ProductReference: FunctionComponent<Props> = props => {
+  const { identifierOption = PRODUCT_REFERENCE_ID, identifierLabel } = props
+  const { item, loading } = useItemContext()
+  const handles = useCssHandles(CSS_HANDLES)
+
+  const [identifierValue, setIdentifierValue] = useState<
+    string | null | undefined
+  >()
+
+  useEffect(() => {
+    if (identifierOption === PRODUCT_REFERENCE_ID) {
+      setIdentifierValue(item.productRefId)
+    } else if (identifierOption === PRODUCT_SKU_REFERENCE_ID) {
+      setIdentifierValue(item.refId)
+    } else if (identifierOption === PRODUCT_SKU_ITEM_ID) {
+      setIdentifierValue(item.id)
+    } else if (identifierOption === PRODUCT_ID) {
+      setIdentifierValue(item.productId)
+    }
+  }, [item.productRefId, identifierOption, item.id, item.productId, item.refId])
+
+  if (loading) {
+    return <Loading />
+  }
```

Comment:
We can replace these `useState` and `useEffect` with a `useMemo`:

```
  const identifierValue = useMemo<string | null | undefined>(() => {
    if (identifierOption === PRODUCT_REFERENCE_ID) return item.productRefId
    if (identifierOption === PRODUCT_SKU_REFERENCE_ID) return item.refId
    if (identifierOption === PRODUCT_SKU_ITEM_ID) return item.id
    if (identifierOption === PRODUCT_ID) return item.productId
  }, [item.productRefId, identifierOption, item.id, item.productId, item.refId])

  if (loading) return <Loading />
  if (!identifierValue) return null
```

Reply:
@kaisermann fixed!

Could you review again, please?